rust/sbp/src/de.rs

```diff
@@ -196,36 +195,91 @@ impl Decoder for SbpDecoder {
             }
         };
         src.advance(start);
-        match parse_frame(src) {
-            Some(Ok(frame)) => Ok(Some(Sbp::from_frame(frame)?)),
-            Some(Err(err)) => Err(err.into()),
-            None => {
-                src.reserve(MAX_FRAME_LEN);
-                Ok(None)
-            }
+        if src.len() < HEADER_LEN {
+            src.reserve(MAX_FRAME_LEN);
+            return Ok(None);
+        }
+        let end = HEADER_LEN + src[PAYLOAD_INDEX] as usize + CRC_LEN;
+        if src.len() < end {
+            src.reserve(MAX_FRAME_LEN);
+            return Ok(None);
         }
+        Ok(Some(Frame(src.split_to(end))))
     }
+}
 
-    fn decode_eof(&mut self, buf: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
-        let res = match self.decode(buf) {
-            Ok(Some(frame)) => Ok(Some(frame)),
-            _ => Ok(None),
-        };
-        buf.clear();
-        res
+/// The wire representation of an SBP message.
+#[derive(Debug)]
+pub struct Frame(BytesMut);
+
+impl Frame {
+    pub fn msg_type(&self) -> u16 {
+        let mut slice = &self.0.chunk()[1..];
+        slice.get_u16_le()
+    }
+
+    pub fn sender_id(&self) -> u16 {
+        let mut slice = &self.0.chunk()[3..];
+        slice.get_u16_le()
+    }
+
+    pub fn payload_len(&self) -> usize {
+        self.0.chunk()[PAYLOAD_INDEX] as usize
+    }
+
+    pub fn payload(&self) -> &[u8] {
+        &self.0.chunk()[HEADER_LEN..HEADER_LEN + self.payload_len()]
+    }
+
+    pub fn crc(&self) -> u16 {
+        let mut slice = &self.0.chunk()[HEADER_LEN + self.payload_len()..];
+        slice.get_u16_le()
+    }
+
+    pub fn check_crc(&self) -> Result<u16, CrcError> {
+        let actual = self.crc();
+        let data = &self.0.chunk()[1..HEADER_LEN + self.payload_len()];
+        if actual == crc16::State::<crc16::XMODEM>::calculate(data) {
+            Ok(actual)
+        } else {
+            Err(CrcError {
+                msg_type: self.msg_type(),
+                sender_id: self.sender_id(),
+                crc: actual,
+            })
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        self.0.chunk().len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    pub fn to_sbp(&self) -> Result<Sbp, Error> {
+        self.check_crc()?;
+        Ok(Sbp::from_field(
+            self.msg_type(),
+            self.sender_id(),
+            self.payload(),
+        )?)
+    }
+
+    pub fn to_bytes(&self) -> Box<dyn AsRef<[u8]> + '_> {
```

Comment:
```
pub fn as_bytes(&self) -> &[u8];
```
`as_` is more common if the conversion is cheap (this is) and you shouldn't need Box dyn here, that more for if you want to return more than one type dynamically

Reply:
says
```
the trait bound `dyn AsRef<[u8]>: std::marker::Sized` is not satisfied [E0277]
```
with AsRef[u8]
instead of Box<dyn AsRef[u8]>

Reply:
nvm fixed using &[u8]